### PR TITLE
add ExponentialFamily and Dirichlet probability distribution

### PR DESCRIPTION
**Pull request by cxxly**
### PR types

New features
### PR changes

APIs
### Describe

- extend Distribution baseclass for supporting multivariant distribution and prob method
- add ExponentialFamily base class and entropy using Bregman divergence
- add Dirichlet probability distribution, including mean, variance, sample, entropy, prob, log_prob method

### Examples
```python
import paddle
# dirichlet distribution
dirichlet = paddle.distribution.Dirichlet(paddle.to_tensor([1., 2., 3.]))
print(dirichlet.entropy())
```

### CODE REVIEW

File: python/paddle/distribution/distribution.py [hunk 1]
```diff
@@ -57,6 +81,14 @@ def kl_divergence(self, other):
         """The KL-divergence between self distributions and other."""
         raise NotImplementedError
 
+    def prob(self, value):
```
Comment (jeff41404) [inline python/paddle/distribution/distribution.py:84]: what's the difference between 'def prob' and 'def probs', why do we need 'def prob' ?

File: python/paddle/distribution/dirichlet.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+#     else:
+#         helper = LayerHelper(op_type, **locals())
+#         out = helper.create_variable_for_type_inference(
+#             dtype=concentration.dtype)
+#         helper.append_op(
+#             type=op_type,
+#             inputs={"Alpha": concentration},
+#             outputs={'Out': out},
+#             attrs={})
+#         return out
```
Comment (jeff41404) [inline python/paddle/distribution/dirichlet.py:162]: these annotations are expected?

File: python/paddle/distribution/distribution.py (hunk 1 above)
Comment (cxxly) [reply python/paddle/distribution/distribution.py:84]: This has been discussed in design document.  Currently, Paddle's pdf and log pdf method named probs and log_prob. It's not consistent, so we add prob method and probs will be deprecated in future.

File: python/paddle/distribution/distribution.py [hunk 3]
```diff
@@ -42,10 +42,34 @@ class Distribution(object):
...
+            Tensor: batch shape
+        """
+        return self._batch_shape
+
+    @property
+    def event_shape(self):
+        """Returns event shape of distribution
+
+        Returns:
+            Tensor: event shape
```
Comment (iclementine) [inline python/paddle/distribution/distribution.py:68]: I don't think the return type of the function is a Tensor.

File: python/paddle/distribution/distribution.py (hunk 1 above)
Comment (iclementine) [reply python/paddle/distribution/distribution.py:84]: can probs() be deprecated?

File: python/paddle/fluid/tests/unittests/distribution/test_distribution_expfamily_static.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+                        self.mock_dist)
+                ])
+
+            np.testing.assert_allclose(
+                out1,
+                out2,
+                rtol=config.RTOL.get(config.DEFAULT_DTYPE),
+                atol=config.ATOL.get(config.DEFAULT_DTYPE))
+
+    def test_entropy_expection(self):
```
Comment (iclementine) [inline python/paddle/fluid/tests/unittests/distribution/test_distribution_expfamily_static.py:55]: exception

File: python/paddle/distribution/exponential_family.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+    def entropy(self):
+        """caculate entropy use `bregman divergence` 
+        https://www.lix.polytechnique.fr/~nielsen/EntropyEF-ICIP2010.pdf
+        """
+        entropy_value = -self._mean_carrier_measure
+
+        natural_parameters = []
+        for parameter in self._natural_parameters:
+            parameter = parameter.detach()
```
Comment (iclementine) [inline python/paddle/distribution/exponential_family.py:44]: detach is not viable in static mode.

Why it is implemented like this?

File: python/paddle/distribution/dirichlet.py (hunk 2 above)
Comment (iclementine) [reply python/paddle/distribution/dirichlet.py:162]: why _dirichlet is commented?

File: python/paddle/fluid/tests/unittests/distribution/test_distribution_expfamily.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+
+@config.place(config.DEVICES)
+@config.parameterize(
+    (config.TEST_CASE_NAME, 'dist'),
+    [('test-dummy-dist', mock.DummyExpFamily(0.5, 0.5)),
+     ('test-dirichlet-dist',
+      paddle.distribution.Dirichlet(paddle.to_tensor(config.xrand())))])
+class TestExponentialFamilyException(unittest.TestCase):
+    def test_entropy_expection(self):
```
Comment (iclementine) [inline python/paddle/fluid/tests/unittests/distribution/test_distribution_expfamily.py:47]: exception

File: python/paddle/distribution/distribution.py (hunk 1 above)
Comment (cxxly) [reply python/paddle/distribution/distribution.py:84]: yes, will be deprecated in future.

File: python/paddle/distribution/dirichlet.py (hunk 2 above)
Comment (cxxly) [reply python/paddle/distribution/dirichlet.py:162]: This pr need other pr dirichlet op https://github.com/PaddlePaddle/Paddle/pull/38244 to be merged, or not unittest will failed.  _dirichlet function will be uncomment in next pr.

File: python/paddle/distribution/distribution.py (hunk 3 above)
Comment (cxxly) [reply python/paddle/distribution/distribution.py:68]: No, it's not a Tensor. It's doc mistake, and will be fixed.

File: python/paddle/fluid/tests/unittests/distribution/test_distribution_expfamily_static.py (hunk 4 above)
Comment (cxxly) [reply python/paddle/fluid/tests/unittests/distribution/test_distribution_expfamily_static.py:55]: will be fixed in next pr.

File: python/paddle/distribution/exponential_family.py (hunk 5 above)
Comment (cxxly) [reply python/paddle/distribution/exponential_family.py:44]: Now it's available in static mode.

File: python/paddle/fluid/tests/unittests/distribution/test_distribution_expfamily.py (hunk 6 above)
Comment (cxxly) [reply python/paddle/fluid/tests/unittests/distribution/test_distribution_expfamily.py:47]: will be fixed.

**Review by iclementine — APPROVED** [review]
LGTM

TODO: fix them in next PR.

**Review by jeff41404 — APPROVED** [review]
LGTM

**Review by TCChenlong — APPROVED** [review]
LGTM
TODO：Fix API Docs

**Review by XiaoguangHu01 — APPROVED** [review]
LG API
